test/parallel/test-stream-some-every.js

```diff
@@ -0,0 +1,87 @@
+'use strict';
+
+const common = require('../common');
+const {
+  Readable,
+} = require('stream');
+const assert = require('assert');
+
+function oneTo5() {
+  return Readable.from([1, 2, 3, 4, 5]);
+}
+
+function oneTo5Async() {
+  return oneTo5().map(async (x) => {
+    await Promise.resolve();
+    return x;
+  });
+}
+{
+  // Some and every work with a synchronous stream and predicate
+  (async () => {
+    assert.strictEqual(await oneTo5().some((x) => x > 3), true);
+    assert.strictEqual(await oneTo5().every((x) => x > 3), false);
+    assert.strictEqual(await oneTo5().some((x) => x > 6), false);
+    assert.strictEqual(await oneTo5().every((x) => x < 6), true);
+    assert.strictEqual(await Readable.from([]).some((x) => true), false);
+    assert.strictEqual(await Readable.from([]).every((x) => true), true);
+  })().then(common.mustCall());
+}
+
+{
+  // Some and every work with an asynchronous stream and synchronous predicate
+  (async () => {
+    assert.strictEqual(await oneTo5Async().some((x) => x > 3), true);
+    assert.strictEqual(await oneTo5Async().every((x) => x > 3), false);
+    assert.strictEqual(await oneTo5Async().some((x) => x > 6), false);
+    assert.strictEqual(await oneTo5Async().every((x) => x < 6), true);
+  })().then(common.mustCall());
+}
+
+{
+  // Some and every works on asynchronous streams with a asynchronous predicate
+  (async () => {
+    assert.strictEqual(await oneTo5().some(async (x) => x > 3), true);
+    assert.strictEqual(await oneTo5().every(async (x) => x > 3), false);
+    assert.strictEqual(await oneTo5().some(async (x) => x > 6), false);
+    assert.strictEqual(await oneTo5().every(async (x) => x < 6), true);
+  })().then(common.mustCall());
+}
+
+{
+  // Short circuiting - some
+  const stream = Readable.from(async function*() {
+    yield 1;
+    yield 2;
+    throw new Error();
+  }());
+  (async () => {
+    assert.strictEqual(await stream.some((x) => x === 2), true);
```

Comment:
nit: you could pass a second parameter to `mustCall` to specify we expect the function called exactly twice.
```suggestion
    assert.strictEqual(await stream.some(mustCall((x) => x === 2), 2), true);
```

Reply:
Isn't that already validated by the error from the stream not being thrown?

Reply:
Not really, the implementation could still call the `fn` function in an unrelated fashion, and that would certainly be a bug that we would want to fix.

Reply:
Isn't that validated in the `.some` and `.every` checks above that check actual functionality?

Reply:
I don't think so, you never test how many calls are made to the function. We could look at it the other way around: is there any reason for not to enforce that those functions are called a specific number of times?

Reply:
Bascially to not test the same thing twice and have exactly one test fail when things break?

Reply:
@aduh95 the test afterwards ("// Short circuiting - every") explicitly checks this behaviour (if more than two calls are made an error is thrown failing the test).

It might make sense to just replace it with the above `common.mustCall` and remove it and test both things at once if you think that's better wdyt?

Reply:
Actually let me just do that that's less code to maintain and still clear enough

Reply:
Yeah the whole test case can be shortened to just:

```js
  await oneTo5().some(common.mustCall((x) => x > 2, 3));
  await oneTo5().every(common.mustCall((x) => x < 3, 3));
```

thanks

Reply:
What I understand from the proposal spec, you should ensure both:

- the `fn` argument is called the exactly 3 times (that's what `mustCall` is testing for).
- the stream is closed immediately after `fn` awaited return value is truthy (or falsy for every) (that's what the `throw new Error` was testing for).

Sorry for being difficult here, but I think there's a nuance, and since it's very cheap to add this test, that's what we should do. If you think both are equivalent, that's maybe because you did the implementation in a way that points are equivalent, but the tests should not be implementation-dependant.

Reply:
> Sorry for being difficult here

I'm not feeling frustrated and in no rush to land this (or any of the other proposal APIs) and I am happy to iterate on this for another month if that's what it takes (really). These sort of API changes suffer from not-enough-eyes more than any other problem.

> the stream is closed immediately after fn awaited return value is truthy (or falsy for every) (that's what the throw new Error was testing for).

That was not my intention I was just testing short-circuit (hence the "short circuit" title in the comment).

We are indeed missing a test like we have in some of the other ones checking whether or not the stream is destroyed after `.every` was called (and ideally one that checks passing an `AbortSignal` that's aborted does not consume the stream probably?)

---

Note since these APIs are experimental I am landing things with a lot less test coverage than I'd normally feel comfortable with and with decisions we may choose to reverse (like `.toArray()` on non-objectMode streams) but I feel like _a lot_ of tests are missing before I'd consider these APIs "reasonably well tested". In particular:

 - we have very few tests that check streams not created from async iterators or `Readable.from` (or streams in push mode in general). 
 - we haven't tested what happens if someone passes a "legacy stream".
 - we haven't tested interop with the version of readable-stream in NPM.
 - we have tested very few error cases (what if you emit an `error` event on the stream while iterating).
 - we haven't tested any "infinite" streams to make sure these methods don't iterate the stream eagerly.
 - we haven't tested concurrency values (other than 1 or 2)
 - we haven't tested what happens when you mess with global `Array` (which usually doesn't matter but when we have APIs where we care about spec alignment it does).

There are a bunch more of these - I expect at least 10-20 more PRs to add coverage and fix bugs in existing APIs before I'd consider these APIs ready for non-experimental use.

Reply:
(I added a test that the stream is destroyed + a check + fix for pre-aborted AbortSignal)

Reply:
very true, glad to hear that improving coverage is on your radar :)